File: web_frontend/backend/routes/extensions.py
from __future__ import annotations

from typing import Any

from fastapi import APIRouter

from web_frontend.backend.runtime_bridge import RuntimeBridge
from web_frontend.backend.routes.utils import optional_package, resource_command


def create_extensions_router(runtime_bridge: RuntimeBridge) -> APIRouter:
    router = APIRouter(prefix="/api/extensions")

    @router.get("")
    async def list_extensions(package_id: str | None = None):
        event = await resource_command(
            runtime_bridge,
            "extensions_manage",
            {"action": "list", **optional_package(package_id)},
            {"extension_configs_listed"},
        )
        return {"event": event}

    @router.post("/mcp")
    async def save_mcp(payload: dict[str, Any]):
        event = await resource_command(
            runtime_bridge,
            "extensions_manage",
            {"action": "upsert_mcp", **payload},
            {"extension_config_updated"},
        )
        return {"event": event}

    @router.post("/mcp/test")
    async def test_mcp(payload: dict[str, Any]):
        event = await resource_command(
            runtime_bridge,
            "extensions_manage",
            {"action": "test_mcp", **payload},
            {"extension_config_tested"},
            timeout_seconds=60.0,
        )
        return {"event": event}

    @router.patch("/mcp/{server_id}")
    async def set_mcp_enabled(server_id: str, payload: dict[str, Any]):
        event = await resource_command(
            runtime_bridge,
            "extensions_manage",
            {
                "action": "set_mcp_enabled",
                "server_id": server_id,
                "enabled": payload.get("enabled", True),
                **optional_package(payload.get("package_id")),
            },
            {"extension_config_updated"},
        )
        return {"event": event}

    @router.delete("/mcp/{server_id}")
    async def remove_mcp(server_id: str, package_id: str | None = None):
        event = await resource_command(
            runtime_bridge,
            "extensions_manage",
            {"action": "remove_mcp", "server_id": server_id, **optional_package(package_id)},
            {"extension_config_updated"},
        )
        return {"event": event}

    @router.post("/skills")
    async def save_skill(payload: dict[str, Any]):
        package_id = payload.get("package_id")
        event = await resource_command(
            runtime_bridge,
            "extensions_manage",
            {
                "action": "upsert_skill",
                "skill": payload.get("skill") if isinstance(payload.get("skill"), dict) else payload,
                "replace_skill_id": payload.get("replace_skill_id"),
                **optional_package(package_id),
            },
            {"extension_config_updated"},
        )
        return {"event": event}

    @router.patch("/skills/{skill_id}")
    async def set_skill_enabled(skill_id: str, payload: dict[str, Any]):
        event = await resource_command(
            runtime_bridge,
            "extensions_manage",
            {
                "action": "set_skill_enabled",
                "skill_id": skill_id,
                "enabled": payload.get("enabled", True),
                **optional_package(payload.get("package_id")),
            },
            {"extension_config_updated"},
        )
        return {"event": event}

    @router.delete("/skills/{skill_id}")
    async def remove_skill(skill_id: str, package_id: str | None = None):
        event = await resource_command(
            runtime_bridge,
            "extensions_manage",
            {"action": "remove_skill", "skill_id": skill_id, **optional_package(package_id)},
            {"extension_config_updated"},
        )
        return {"event": event}

    return router

File: web_frontend/frontend/src/api/extensions.ts
import type { ToolPermissionOverrideView, ToolPermissionPolicyView } from '@/types/protocol'
import type { McpServerConfig, SkillConfig, WorkspaceContextInput, WorkspaceRequestContext } from './resourceTypes'
import { requestEvent, withQuery } from './http'

type ExtensionContext = Pick<WorkspaceRequestContext, 'packageId' | 'resourceMode'>

export const extensionsApi = {
  list: (context?: WorkspaceContextInput) => requestEvent(withQuery('/api/extensions', extensionContextQuery(context))),
  saveMcp: (server: McpServerConfig, context?: WorkspaceContextInput) =>
    requestEvent('/api/extensions/mcp', {
      method: 'POST',
      body: JSON.stringify({ server, ...extensionContextPayload(context) }),
    }),
  testMcp: (serverIdOrConfig: string | McpServerConfig, context?: WorkspaceContextInput) => {
    const payload =
      typeof serverIdOrConfig === 'string'
        ? { server_id: serverIdOrConfig, ...extensionContextPayload(context) }
        : { server: serverIdOrConfig, ...extensionContextPayload(context) }
    return requestEvent('/api/extensions/mcp/test', {
      method: 'POST',
      body: JSON.stringify(payload),
    })
  },
  setMcpEnabled: (serverId: string, enabled: boolean, context?: WorkspaceContextInput) =>
    requestEvent(`/api/extensions/mcp/${encodeURIComponent(serverId)}`, {
      method: 'PATCH',
      body: JSON.stringify({ enabled, ...extensionContextPayload(context) }),
    }),
  removeMcp: (serverId: string, context?: WorkspaceContextInput) =>
    requestEvent(withQuery(`/api/extensions/mcp/${encodeURIComponent(serverId)}`, extensionContextQuery(context)), {
      method: 'DELETE',
    }),
  saveSkill: (skill: SkillConfig, context?: WorkspaceContextInput) =>
    requestEvent('/api/extensions/skills', {
      method: 'POST',
      body: JSON.stringify({
        skill,
        replace_skill_id: skill.replace_skill_id,
        ...extensionContextPayload(context),
      }),
    }),
  setSkillEnabled: (skillId: string, enabled: boolean, context?: WorkspaceContextInput) =>
    requestEvent(`/api/extensions/skills/${encodeURIComponent(skillId)}`, {
      method: 'PATCH',
      body: JSON.stringify({ enabled, ...extensionContextPayload(context) }),
    }),
  removeSkill: (skillId: string, context?: WorkspaceContextInput) =>
    requestEvent(withQuery(`/api/extensions/skills/${encodeURIComponent(skillId)}`, extensionContextQuery(context)), {
      method: 'DELETE',
    }),
  updateToolPermissions: (policy: ToolPermissionPolicyView, context?: WorkspaceContextInput) =>
    requestEvent('/api/extensions/tool-permissions', {
      method: 'PUT',
      body: JSON.stringify({ policy, ...extensionContextPayload(context) }),
    }),
  setToolPermission: (toolId: string, override: ToolPermissionOverrideView, context?: WorkspaceContextInput) =>
    requestEvent(`/api/extensions/tool-permissions/${encodeURIComponent(toolId)}`, {
      method: 'PATCH',
      body: JSON.stringify({ override, ...extensionContextPayload(context) }),
    }),
  resetToolPermission: (toolId: string, context?: WorkspaceContextInput) =>
    requestEvent(withQuery(`/api/extensions/tool-permissions/${encodeURIComponent(toolId)}`, extensionContextQuery(context)), {
      method: 'DELETE',
    }),
}

function extensionContextPayload(context?: WorkspaceContextInput): { package_id?: string; resource_mode?: string } {
  const normalized = normalizeExtensionContext(context)
  return {
    ...(normalized.packageId ? { package_id: normalized.packageId } : {}),
    ...(normalized.resourceMode ? { resource_mode: normalized.resourceMode } : {}),
  }
}

function extensionContextQuery(context?: WorkspaceContextInput): { package_id?: string; resource_mode?: string } {
  return extensionContextPayload(context)
}

function normalizeExtensionContext(context?: WorkspaceContextInput): ExtensionContext {
  if (typeof context === 'string') return { packageId: context || undefined }
  if (!context) return {}
  return {
    packageId: context.packageId || undefined,
    resourceMode: context.resourceMode,
  }
}
